platform/cms/accelerated_publishing/rfc/broken-link-checking.md

```diff
@@ -0,0 +1,126 @@
+
+1. Moving from Content Build to Next Build will happen incrementally. There will be a period during which both systems will be responsible for portions of CMS content. This presents problems for points 2. and 3. above. Content Build _requires_ building the entire set of CMS pages in order to do its link checks; but, Content Build will no longer have the entire set of CMS pages once we start shifting some content to Next Build.
+2. The speed gains of Next Build / Accelerated Publishing will come from "incremental publishing". Put simply, Content Build operates by publishing every single CMS page every single publishing run. Most of this publishing is not useful, because nothing has changed in the content. With Next Build, we will transition to a paradigm where we only build pages that have changes. This should drastically reduce the amount of pages that a single publishing run requires. However this also means these pages will be built in small batches. This also presents problems for points 2. and 3. above.
+3. Finally, link checking can be a time-consuming process. With our stated goal of increasing publishing speed and being able to expand the amount of content the CMS can maintain, we want to avoid adding time to publishing. Rather than checking links at build time and having it add to that build time, it would be good to test for broken links independently.
+
+### Problems to solve
+
+In summary, the problems that the current system has are:
+
+- Content Build will no longer have complete knowledge of all pages, and so any pages that have migrated to Next Build will be flagged as broken links by Content Build
+- We will still have to test for broken links for the site as a whole.
+- Running full broken links checks at build time is too slow for an accelerated publishing scenario.
+
+## Design
+
+_Explain the proposed changes in enough detail so that a team member with working knowledge
+of the project could implement the change themselves. Bonus points if this is easily mappable
+to user stories or specs._
+
+### Link Checking Features
+
+Our system must do the following
+
+- It must operate on the entirety of va.gov content, regardless of what system the page was built by.
+- It must not create an impediment to the build process.
+- It must report the broken links into a central location that can respond to the issues.
+
+### Crawler Proposal
+
+A straight-forward way would be to create a site crawler that checks va.gov pages, follows any links found, and reports any that return a non-200/300 response. This would operate as a CI process on VA.gov's published content.
+
+All content published by Content Build and Next Build is listed in sitemap.xml files. These URLs can be read by a site crawler and the contents of the URL scanned.
+
+Advantages:
+
+- This would be relatively simple to create.
```

Comment:
Checking links from within the VA network is not as simple as it sounds.  Bumped my head into this many times.  Checking links to other gov't sites is not as simple as it sounds either.   Many of them return responses that lie.

Reply:
Fair point; though, for prod at least, this is likely to run outside the network. We're concerned that a Veteran can reach these pages, not that they're accessible from within the VA network, so it would actually be prudent to *not* run the scan from within the VA network.

"Gov't sites that give bad responses" is noted though; can you speak more to specifics?

Reply:
There are a lot of site that when you try to curl them return something other than a friendly 200 response.  [Most recent example](https://dsva.slack.com/archives/C02730UEZPS/p1702659644868359)

So the method has to build in an ignore list  similar to the config we have in node_link_report  
![image](https://github.com/department-of-veterans-affairs/va.gov-team/assets/5752113/dfcbc3de-16af-403f-960b-8fbe33bcbf14)

In fact it might make sense to have it use the same ignore list so that the cms and the FE are in agreement about what they are ignoring.

Reply:
We could. We'd have to figure out how to make the CMS list available more publicly, so that it's available to an out-of-VA-network process.

I'd also really like to explore if we can figure out what leads the site to return a 403 to a CURL process but not to a browser request. At the end of the day both are just making HTTP calls with different headers, or at least that's my naive understanding. If we can figure out what allows a browser request to succeed, maybe we can also automate checks to those same domains.

I understand why the domain list is necessary, but it frankly worries me a bit because of not being able to differentiate legitimate 404s/403s from 'unfriendly' ones.

Reply:
See that is the beauty of having a public repo.  The data is right there in readable yml format :smiley: 
https://github.com/department-of-veterans-affairs/va.gov-cms/blob/main/config/sync/node_link_report.settings.yml#L7-L98   


Why do sites get goofy.... too many reasons.  Could we masquerade as a browser user agent... we could.  Not really supposed to, but we could.  Though even when I masquerade as chrome, that one example site still gives me a 403.

Reply:
Though in fairness, not all of those need ignoring. Some are only ignored because of the TIC intercepting, while the others are truly bad responses.

Reply:
Right, and again I think it's much better to be checking this stuff from outside the TIC. I'd be curious to see if we get the same kinds of misbehaving responses from these sites if the TIC is not contributing. 

If the purpose of our scans is to check that a browser can hit a given site, I don't think it's a problem to modify user agent. We're not scraping or spidering except to verify the site works when hit with a browser.